File: Easy/JAVA/SortedSquaredArray.java
package JAVA;

import java.util.Arrays;

class SortedSquaredArray {
    // O(nlogn) time | O(n) space
    public static int[] sortedSquaredArray1(int[] array) {
        int[] sorted = new int[array.length];
        for (int i = 0; i < array.length; i++) {
            sorted[i] = (int) Math.pow(array[i], 2);
        }
        Arrays.sort(sorted);

        return sorted;
    }

    // O(n) time | O(n) space
    public static int[] sortedSquaredArray2(int[] array) {
        int[] sorted = new int[array.length];
        int start = 0;
        int end = array.length - 1;
        for (int i = array.length - 1; i >= 0; i--) {
            if (array[start] < array[end]) {
                sorted[i] = (int) Math.pow(array[end], 2);
                end--;
            } else {
                sorted[i] = (int) Math.pow(array[start], 2);
                start++;
            }
        }

        return sorted;
    }

    public static void main(String[] args) {
        int[] array = new int[] { 1, 2, 3, 5, 6, 8, 9 };
        int[] result = sortedSquaredArray2(array);
        System.out.println(Arrays.toString(result));
    }
}

File: Easy/JAVASCRIPT/SortedSquaredArray.js
// O(n log(n)) | O(n) space
sortedSquaredArray1 = (array) => {
    const sorted = [];
    for (let i = 0; i < array.length; i++) {
        sorted[i] = Math.pow(array[i], 2);
    }
    sorted.sort((num1, num2) => { num1 - num2 });
    return sorted;
}

// O(n) time | O(n) space
sortedSquaredArray2 = (array) => {
    const sorted = [];
    let start = 0;
    let end = array.length - 1;
    for (let i = array.length - 1; i >= 0; i--) {
        if (array[start] < array[end]) {
            sorted[i] = array[end];
            end--;
        } else {
            sorted[i] = array[start];
            start++;
        }
    }

    return sorted;
}

const result = sortedSquaredArray1([1, 2, 3, 5, 6, 8, 9]);
console.log(result);